include/linker/linker-tool-gcc.h

```diff
@@ -111,7 +111,7 @@
  */
 #if defined(CONFIG_ARCH_POSIX)
 #define GROUP_ROM_LINK_IN(vregion, lregion)
-#elif defined(Z_VM_KERNEL)
+#elif defined(CONFIG_XIP) || defined(Z_VM_KERNEL)
```

Comment:
@dcpleung @andyross This is the only way I was able to have `common-rom.ld` content properly mapped into ESP virtual memory. Otherwise all those k_objects hardcoded in RODATA section needs to be kept.  
In case this is not the right approach, any recommendation is welcome.

Reply:
I am worried that this change will affect other XIP enabled platforms since the linking will be done differently.

IIRC, when virtual memory is first introduced, it is not compatible with XIP, and there was never an effort to make them work together (yet). The issue here is that both of these require the same linker magic (like the line here) to work.

And it seems your later commits are disabling CONFIG_XIP, so why do you still need this?

Reply:
@dcpleung, Later commits are not disabling XIP. In fact, it is setting enabled by default in Kconfig.soc. `select XIP`.  
I am also afraid about the behavior on other platforms.
I just found this:  
```c
#if Z_MEM_VM_OFFSET != 0
#define Z_VM_KERNEL 1
#ifdef CONFIG_XIP
#error "XIP and a virtual memory kernel are not allowed"
#endif
#endif
```
What would be the alternatives?  
a) Having custom ESP check in that file to enable this?
b) Some different approach in ESP linker script? (I could not find a way for this though).

@nashif Do you have something to add?

Reply:
One way I can think of is not to enable XIP, and build with virtual memory addresses. However, some XIP-like magic will need to be introduced into your reset code to copy the data and BSS regions into memory. Or if you have a separate bootloader, do the loading there, and Zephyr kernel does not have to worry about copying.

Reply:
@dcpleung @andyross , boot up code already does that, but we still need both LMA and VMA, which is missing for current ROM sections (which is intended in this commit). So if boot code is supposed to do that mapping when XIP is enabled, then this change is correct.  
Also, `GROUP_DATA_LINK_IN` has the same check added in this commit:
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/linker/linker-tool-gcc.h#L134  
Make sense having `GROUP_ROM_LINK_IN` to have as well.

Reply:
Since you are using virtual memory, I think it would be better to set `Z_VM_KERNEL`. Change `#ifdef CONFIG_MMU` to `#ifdef CONFIG_KERNEL_VM_SUPPORT` would probably work in `include/sys/mem_manage.h`.

Reply:
Well, it did not work using this approach.  
Have you checked my previous comment regarding `GROUP_DATA_LINK_IN` having this XIP check?

Reply:
`GROUP_DATA_LINK_IN` is for data region which is supposed to be copied to RAM if XIP is enabled, hence the need for load address. RODATA is NOT supposed to be copied to RAM. If you change the #if here, it will definitely affect other XIP boards as now they will have a load address for RODATA where previously not.

To avoid affecting other boards, it is be possible to undefine `GROUP_ROM_LINK_IN` in your linker script so you can redefine it to suit your need.

Reply:
But redefining `GROUP_ROM_LINK_IN` in the linker won't handle the case  as in this test, will it?  
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/misc/iterable_sections/sections-rom.ld

Reply:
Closing this conversation as this approach isn't valid.